File: crates/pulsing-workspace/src/journal.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{bail, Context, Result};
use chrono::Utc;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use walkdir::WalkDir;

use crate::ignore::should_skip;
use crate::layout::WorkspaceLayout;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileRecord {
    pub path: String,
    pub sha256: String,
    pub size: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RevisionManifest {
    pub id: String,
    pub parent: Option<String>,
    pub created_at: String,
    pub message: String,
    pub author: String,
    pub files: Vec<FileRecord>,
}

#[derive(Debug, Clone)]
pub struct RevisionInfo {
    pub id: String,
    pub created_at: String,
    pub message: String,
    pub author: String,
    pub file_count: usize,
}

#[derive(Debug, Clone)]
pub struct CheckpointOptions {
    pub message: Option<String>,
    pub author: Option<String>,
}

#[derive(Debug, Clone)]
pub struct RollbackOptions {
    pub revision_id: Option<String>,
}

pub fn list_revisions(layout: &WorkspaceLayout) -> Result<Vec<RevisionInfo>> {
    let mut out = Vec::new();
    let rev_dir = layout.revisions_dir();
    if !rev_dir.is_dir() {
        return Ok(out);
    }
    for entry in fs::read_dir(&rev_dir)? {
        let entry = entry?;
        if !entry.file_type()?.is_dir() {
            continue;
        }
        let manifest_path = entry.path().join("manifest.json");
        if !manifest_path.is_file() {
            continue;
        }
        let manifest: RevisionManifest =
            serde_json::from_str(&fs::read_to_string(&manifest_path)?)?;
        out.push(RevisionInfo {
            id: manifest.id,
            created_at: manifest.created_at,
            message: manifest.message,
            author: manifest.author,
            file_count: manifest.files.len(),
        });
    }
    out.sort_by(|a, b| a.id.cmp(&b.id));
    Ok(out)
}

pub fn current_head(layout: &WorkspaceLayout) -> Result<Option<String>> {
    let head = layout.head_file();
    if !head.is_file() {
        return Ok(None);
    }
    let id = fs::read_to_string(&head)?.trim().to_string();
    if id.is_empty() {
        Ok(None)
    } else {
        Ok(Some(id))
    }
}

pub fn checkpoint(layout: &WorkspaceLayout, opts: CheckpointOptions) -> Result<RevisionManifest> {
    let parent = current_head(layout)?;
    let next_id = next_revision_id(layout)?;
    let rev_path = layout.revision_dir(&next_id);
    let files_dir = rev_path.join("files");
    fs::create_dir_all(&files_dir)?;

    let scanned = scan_workspace_files(layout)?;
    let mut records = Vec::new();
    for rel in scanned {
        let src = layout.root.join(&rel);
        let dest = files_dir.join(&rel);
        if let Some(parent) = dest.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::copy(&src, &dest)?;
        let data = fs::read(&src)?;
        let mut hasher = Sha256::new();
        hasher.update(&data);
        let hash = format!("{:x}", hasher.finalize());
        records.push(FileRecord {
            path: rel.to_string_lossy().into_owned(),
            sha256: hash,
            size: data.len() as u64,
        });
    }

    let manifest = RevisionManifest {
        id: next_id.clone(),
        parent,
        created_at: Utc::now().to_rfc3339(),
        message: opts
            .message
            .unwrap_or_else(|| "checkpoint".to_string()),
        author: opts.author.unwrap_or_else(|| "user".to_string()),
        files: records,
    };

    fs::write(
        rev_path.join("manifest.json"),
        serde_json::to_string_pretty(&manifest)? + "\n",
    )?;
    fs::write(layout.head_file(), format!("{next_id}\n"))?;
    Ok(manifest)
}

pub fn rollback(layout: &WorkspaceLayout, opts: RollbackOptions) -> Result<RevisionManifest> {
    let id = match opts.revision_id {
        Some(id) => id,
        None => current_head(layout)?.context("no checkpoint to roll back to")?,
    };
    let rev_path = layout.revision_dir(&id);
    let manifest_path = rev_path.join("manifest.json");
    if !manifest_path.is_file() {
        bail!("revision {id} not found");
    }
    let manifest: RevisionManifest =
        serde_json::from_str(&fs::read_to_string(&manifest_path)?)?;
    let files_dir = rev_path.join("files");

    for file in &manifest.files {
        let rel = Path::new(&file.path);
        if should_skip(rel) {
            continue;
        }
        let src = files_dir.join(rel);
        let dest = layout.root.join(rel);
        if !src.is_file() {
            continue;
        }
        if let Some(parent) = dest.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::copy(&src, &dest)?;
    }

    fs::write(layout.head_file(), format!("{id}\n"))?;
    Ok(manifest)
}

fn next_revision_id(layout: &WorkspaceLayout) -> Result<String> {
    let existing = list_revisions(layout)?;
    let next = existing
        .last()
        .map(|r| r.id.parse::<u32>().unwrap_or(0) + 1)
        .unwrap_or(1);
    Ok(format!("{next:04}"))
}

fn scan_workspace_files(layout: &WorkspaceLayout) -> Result<Vec<PathBuf>> {
    let mut paths = Vec::new();
    for entry in WalkDir::new(&layout.root)
        .follow_links(false)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        if !entry.file_type().is_file() {
            continue;
        }
        let path = entry.path();
        let rel = layout
            .rel_to_root(path)
            .with_context(|| format!("path outside workspace: {}", path.display()))?;
        if should_skip(&rel) {
            continue;
        }
        paths.push(rel);
    }
    paths.sort();
    Ok(paths)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::init::{init_workspace, InitOptions, Template};
    use tempfile::tempdir;

    #[test]
    fn init_checkpoint_rollback_roundtrip() {
        let dir = tempdir().unwrap();
        let root = dir.path();
        fs::write(root.join("hello.txt"), b"v1").unwrap();

        init_workspace(
            root,
            InitOptions {
                template: Template::Minimal,
                name: None,
                force: false,
            },
        )
        .unwrap();

        let layout = WorkspaceLayout::new(root);
        fs::write(root.join("hello.txt"), b"v2").unwrap();
        checkpoint(
            &layout,
            CheckpointOptions {
                message: Some("v2".into()),
                author: None,
            },
        )
        .unwrap();

        fs::write(root.join("hello.txt"), b"v3").unwrap();
        rollback(&layout, RollbackOptions { revision_id: None }).unwrap();
        assert_eq!(fs::read(root.join("hello.txt")).unwrap(), b"v2");
    }
}

File: python/pulsing/workspace/journal.py
# SPDX-License-Identifier: Apache-2.0
"""Workspace journal: checkpoint and rollback."""

from __future__ import annotations

import hashlib
import json
import shutil
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from pulsing.workspace.ignore import should_skip
from pulsing.workspace.layout import WorkspaceLayout


@dataclass
class RevisionInfo:
    id: str
    created_at: str
    message: str
    author: str
    file_count: int


def _scan_files(layout: WorkspaceLayout) -> list[Path]:
    paths: list[Path] = []
    for path in layout.root.rglob("*"):
        if not path.is_file():
            continue
        rel = path.relative_to(layout.root)
        if should_skip(rel):
            continue
        paths.append(rel)
    paths.sort()
    return paths


def current_head(layout: WorkspaceLayout) -> str | None:
    if not layout.head_file.is_file():
        return None
    value = layout.head_file.read_text(encoding="utf-8").strip()
    return value or None


def list_revisions(layout: WorkspaceLayout) -> list[RevisionInfo]:
    out: list[RevisionInfo] = []
    if not layout.revisions_dir.is_dir():
        return out
    for entry in sorted(layout.revisions_dir.iterdir()):
        if not entry.is_dir():
            continue
        manifest_path = entry / "manifest.json"
        if not manifest_path.is_file():
            continue
        data = json.loads(manifest_path.read_text(encoding="utf-8"))
        out.append(
            RevisionInfo(
                id=str(data["id"]),
                created_at=str(data["created_at"]),
                message=str(data.get("message", "")),
                author=str(data.get("author", "")),
                file_count=len(data.get("files") or []),
            )
        )
    return out


def _next_revision_id(layout: WorkspaceLayout) -> str:
    revs = list_revisions(layout)
    if not revs:
        return "0001"
    last = int(revs[-1].id)
    return f"{last + 1:04}"


def checkpoint(
    layout: WorkspaceLayout,
    *,
    message: str | None = None,
    author: str | None = None,
) -> dict[str, Any]:
    from pulsing.workspace.hooks import run_before_checkpoint, run_after_checkpoint

    ctx = {
        "root": str(layout.root),
        "message": message or "checkpoint",
    }
    extra = run_before_checkpoint(ctx)

    rev_id = _next_revision_id(layout)
    rev_path = layout.revision_dir(rev_id)
    files_dir = rev_path / "files"
    files_dir.mkdir(parents=True, exist_ok=True)

    scanned = _scan_files(layout)
    if extra:
        for rel_str in extra:
            rel = Path(rel_str)
            p = layout.root / rel
            if p.is_file() and rel not in scanned:
                scanned.append(rel)
        scanned.sort()

    records: list[dict[str, Any]] = []
    for rel in scanned:
        src = layout.root / rel
        dest = files_dir / rel
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dest)
        data = src.read_bytes()
        records.append(
            {
                "path": rel.as_posix(),
                "sha256": hashlib.sha256(data).hexdigest(),
                "size": len(data),
            }
        )

    manifest: dict[str, Any] = {
        "id": rev_id,
        "parent": current_head(layout),
        "created_at": datetime.now(timezone.utc).isoformat(),
        "message": message or "checkpoint",
        "author": author or "user",
        "files": records,
    }
    (rev_path / "manifest.json").write_text(
        json.dumps(manifest, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    layout.head_file.parent.mkdir(parents=True, exist_ok=True)
    layout.head_file.write_text(f"{rev_id}\n", encoding="utf-8")
    run_after_checkpoint({**ctx, "revision_id": rev_id})
    return manifest


def rollback(
    layout: WorkspaceLayout, *, revision_id: str | None = None
) -> dict[str, Any]:
    rev_id = revision_id or current_head(layout)
    if not rev_id:
        raise SystemExit("no checkpoint to roll back to")
    rev_path = layout.revision_dir(rev_id)
    manifest_path = rev_path / "manifest.json"
    if not manifest_path.is_file():
        raise SystemExit(f"revision {rev_id} not found")
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    files_dir = rev_path / "files"
    for file in manifest.get("files") or []:
        rel = Path(file["path"])
        if should_skip(rel):
            continue
        src = files_dir / rel
        dest = layout.root / rel
        if not src.is_file():
            continue
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dest)
    layout.head_file.write_text(f"{rev_id}\n", encoding="utf-8")
    return manifest
